client/components/keyed-suggestions/trigrams.js

```diff
@@ -0,0 +1,63 @@
+export const trigrams = ( str ) => {
+	const n = 3;
+	const grams = [];
+	const input = [ '_BEGIN_', ...str.split( '' ), '_END_' ];
+	for ( let i = 0; i <= input.length - n; i++ ) {
+		grams.push( input.slice( i, i + n ).join( '' ) );
+	}
+	return grams;
+};
+
+export const grams_to_lookup = ( gram_list ) => {
+	const lookup = {};
+	for ( const gram of gram_list ) {
+		if ( gram in lookup ) {
+			lookup[ gram ] += 1;
+		} else {
+			lookup[ gram ] = 1;
+		}
+	}
+	return lookup;
+};
+
+export const lookup_to_magnitude = ( lookup ) => {
+	if ( ! lookup || Object.keys( lookup ).length === 0 ) {
+		return 0;
+	}
+	return Math.sqrt(
+		Object.values( lookup )
+			.map( ( x ) => x * x )
+			.reduce( ( a, b ) => a + b )
+	);
+};
+
+const string_to_lookup_cache = {};
+
+export const string_to_lookup = ( str ) => {
```

Comment:
Nitpick: I personally prefer not to export every function unless it's necessary since: When a function is NOT exported, it's absolutely clear that it's only used in one place. If I know a function is not used elsewhere (since it's not even exported), I can refactor it without checking where else it is used.